src/hotspot/share/cds/aotClassLinker.cpp

```diff
@@ -0,0 +1,238 @@
+    if (!try_add_candidate(intf)) {
+      return false;
+    }
+  }
+
+  add_candidate(ik);
+
+  if (log_is_enabled(Info, cds, aot, link)) {
+    ResourceMark rm;
+    log_info(cds, aot, link)("%s %s", ArchiveUtils::class_category(ik), ik->external_name());
+  }
+
+  return true;
+}
+
+void AOTClassLinker::add_candidates() {
+  assert_at_safepoint();
+  if (CDSConfig::is_dumping_aot_linked_classes()) {
+    GrowableArray<Klass*>* klasses = ArchiveBuilder::current()->klasses();
+    for (GrowableArrayIterator<Klass*> it = klasses->begin(); it != klasses->end(); ++it) {
+      Klass* k = *it;
+      if (k->is_instance_klass()) {
+        try_add_candidate(InstanceKlass::cast(k));
+      }
+    }
+  }
+}
+
+void AOTClassLinker::write_to_archive() {
+  assert(is_initialized(), "sanity");
+  assert_at_safepoint();
+
+  if (CDSConfig::is_dumping_aot_linked_classes()) {
+    AOTLinkedClassTable* table = AOTLinkedClassTable::get(CDSConfig::is_dumping_static_archive());
+    table->set_boot(write_classes(nullptr, true));
+    table->set_boot2(write_classes(nullptr, false));
+    table->set_platform(write_classes(SystemDictionary::java_platform_loader(), false));
+    table->set_app(write_classes(SystemDictionary::java_system_loader(), false));
+  }
+}
+
+Array<InstanceKlass*>* AOTClassLinker::write_classes(oop class_loader, bool is_javabase) {
+  ResourceMark rm;
+  GrowableArray<InstanceKlass*> list;
+
+  for (int i = 0; i < _sorted_candidates->length(); i++) {
+    InstanceKlass* ik = _sorted_candidates->at(i);
+    if (ik->class_loader() != class_loader) {
+      continue;
+    }
+    if ((ik->module() == ModuleEntryTable::javabase_moduleEntry()) != is_javabase) {
+      continue;
+    }
+
+    if (ik->is_shared() && CDSConfig::is_dumping_dynamic_archive()) {
+      if (CDSConfig::is_using_aot_linked_classes()) {
+        // This class was recorded as a AOT-linked for the base archive,
+        // so there's no need to do so again for the dynamic archive.
+      } else {
+        list.append(ik);
+      }
+    } else {
+      list.append(ArchiveBuilder::current()->get_buffered_addr(ik));
+    }
+  }
+
+  if (list.length() == 0) {
+    return nullptr;
+  } else {
+    const char* category = ArchiveUtils::class_category(list.at(0));
+    log_info(cds, aot, link)("written %d class(es) for category %s", list.length(), category);
+    return ArchiveUtils::archive_array(&list);
+  }
+}
+
+int AOTClassLinker::num_platform_initiated_classes() {
+  // AOTLinkedClassBulkLoader will initiate loading of all public boot classes in the platform loader.
+  return num_initiated_classes(nullptr, nullptr);
+}
+
+int AOTClassLinker::num_app_initiated_classes() {
+  // AOTLinkedClassBulkLoader will initiate loading of all public boot/platform classes in the app loader.
+  return num_initiated_classes(nullptr, SystemDictionary::java_platform_loader());
+}
+
+int AOTClassLinker::num_initiated_classes(oop loader1, oop loader2) {
```

Comment:
The two loader arguments here are quite confusing marking it hard to understand the code. Can it be refactored as this:

```
int AOTClassLinker::num_platform_initiated_classes() {
  // AOTLinkedClassBulkLoader will initiate loading of all public boot classes in the platform loader.
  return num_initiated_classes(nullptr);
}

int AOTClassLinker::num_app_initiated_classes() {
  // AOTLinkedClassBulkLoader will initiate loading of all public boot/platform classes in the app loader.
  return num_platform_initiated_classes + num_initiated_classes(SystemDictionary::java_platform_loader());
}

int AOTClassLinker::num_initiated_classes(oop loader) {
  int n = 0;
  for (int i = 0; i < _sorted_candidates->length(); i++) {
    InstanceKlass* ik = _sorted_candidates->at(i);
    if (ik->is_public() && !ik->is_hidden() &&
        (ik->class_loader() == loader) {
      n++;
    }
  }

  return n;
}

Reply:
I renamed the function to `AOTClassLinker::count_public_classes()` and made it handle only a single loader each time. This function is not speed critical so I think this way it's much easier to read.